### DISCUSSION TITLE
Converting the overnight hypnogram and handling abnormal spectrogram values

### DISCUSSION THREAD

**Discussion by qaqis95** [raphaelvallat/yasa#143] [Q&A] [answered]
Excellent work!

1. Extracting sleep EEG data from the sleep EDF file and converting it to the npz format, I obtained very low values for the first five elements. What could be the cause of this outcome?
 F4-M1: [-3.05180438e-08 -3.05180438e-08 -3.05180438e-08 -3.05180438e-08
 -3.05180438e-08]
C4-M1: [-3.05180438e-08 -3.05180438e-08 -3.05180438e-08 -3.05180438e-08
 -3.05180438e-08]
O2-M1: [-3.05180438e-08 -3.05180438e-08 -3.05180438e-08 -3.05180438e-08
 -3.05180438e-08]
F3-M2: [-3.05180438e-08 -3.05180438e-08 -3.05180438e-08 -3.05180438e-08
 -3.05180438e-08]
C3-M2: [-3.05180438e-08 -3.05180438e-08 -3.05180438e-08 -3.05180438e-08
 -3.05180438e-08]
O1-M2: [-3.05180438e-08 -3.05180438e-08 -3.05180438e-08 -3.05180438e-08
 -3.05180438e-08]

2. The spectrogram values obtained from my overnight sleep data are lower compared to the example data. This raises the question of what could be causing this discrepancy.
<img width="755" alt="iShot2023-05-28 17 16 47" src="https://github.com/raphaelvallat/yasa/assets/134837471/fa888d9e-3701-4a1a-9e4b-496580c2a079">

3. I would like to replicate the "Artifact Rejection" section. 
I initially converted the overnight hypnogram from txt format to npz format, which resulted in 1308 arrays. However, when converting the entire overnight EEG to npz format, it resulted in 10049536 arrays, causing a data mismatch. How can this issue be resolved?
Below is my code.

Stage2 = np.loadtxt('2.txt')
print(Stage2)
np.savez('Stage2.npz', data=Stage2)
f = np.load('data.npz')
data= f['data']
chan = ['F4-M1', 'C4-M1', 'O2-M1', 'F3-M2', 'C3-M2', 'O1-M2']
sf = 256.
hypno = np.load('Stage2.npz').get('data')
times = np.arange(data.size) / sf
print(data.shape)
print(chan)
print(np.round(data[:, 0:5], 3))
data.shape, hypno.shape

---output--
[0. 0. 0. ... 0. 0. 0.]
(6, 10049536)
['F4-M1', 'C4-M1', 'O2-M1', 'F3-M2', 'C3-M2', 'O1-M2']
[[-0. -0. -0. -0. -0.]
 [-0. -0. -0. -0. -0.]
 [-0. -0. -0. -0. -0.]
 [-0. -0. -0. -0. -0.]
 [-0. -0. -0. -0. -0.]
 [-0. -0. -0. -0. -0.]]
((6, 10049536), (1308,))

**Comment by raphaelvallat** [comment, marked as answer]
Hi,

1. Are you using MNE to load your EDF data? MNE automatically converts the data to Volts, so you need to multiply by 1e6 if you want the data to be in micro-Volts.

2. Do you mean lower in terms of the log-power unit, in dB / Hz? The absolute amplitude of the EEG is impacted by age, sex, skull thickness, as well as the PSG system and electrode location, so  would not worry too much about this.

3. You need to upsample your hypnogram to the sampling frequency of the EEG data, by using this function: https://raphaelvallat.com/yasa/build/html/generated/yasa.hypno_upsample_to_data.html as explained in [this section](https://raphaelvallat.com/yasa/build/html/quickstart.html#id10) of the quickstart

**Reply by qaqis95** [reply]
Thanks for your reply @raphaelvallat . It solved my issue!

New questions are about the SO-spindle function.

1.  After running the following code, I will obtain 17 variables:
Variable names in sw.summary(): ['Start', 'NegPeak', 'MidCrossing', 'PosPeak', 'End', 'Duration', 'ValNegPeak', 'ValPosPeak', 'PTP', 'Slope', 'Frequency', 'Stage', 'Channel', 'IdxChannel', 'CooccurringSpindle', 'CooccurringSpindlePeak', 'DistanceSpindleToSW']

sw.find_cooccurring_spindles(sp.summary(), lookaround=1.2)
print(sw.summary().round(3))
variable_names = list(sw.summary().columns)
print("Variable names in sw.summary():", variable_names)

![1686067634503](https://github.com/raphaelvallat/yasa/assets/134837471/2711d526-01d5-4ba5-89cc-f72e46a2e20a)

However, when I extract the So-spindle coupling, 3 columns ('CooccurringSpindle', 'CooccurringSpindlePeak', 'DistanceSpindleToSW') are missing. 
![1686067770836](https://github.com/raphaelvallat/yasa/assets/134837471/77dc52de-8f8c-43ef-8f71-a2b7aa1264f1)

I want to obtain the strength and phase of SSO coupling. How can this issue be resolved?

2. In this subject, there are a total of 2880 spindles and 6683 SO, but only 78 so-spindle coupling. Is this normal?

**Reply by raphaelvallat** [reply]
@qaqis95 regarding the SO-spindle coupling, I would not recommend using the `find_cooccurring_spindles` function to determine the coupling. Instead, you can just use `yasa.sw_detect(coupling=True)`, which will calculate, for each slow-wave, the coupling with the filtered sigma band. This is what most people use in publications. 

If you want to keep using the `find_cooccurring_spindles` function, I would recommend using more liberal thresholds for the spindle detection, in order to include more spindles.

Regarding the missing columns, I think it's because everytime you are recalculating `sw.summary()`, it starts from scratch and you would need to re-use the `find_cooccurring_spindles` function.

Hope this helps
Raphael

**Reply by qaqis95** [reply]
Thank you @raphaelvallat once again for your reply! Very helpful！

I used the 'find_cooccurring_spindles' because in some individuals, the number of so-spindle couplings I obtained on the same channel was much larger than the number of spindles. I felt that there might be an issue, and after seeing your previous responses, I decided to try the 'find_cooccurring_spindles' function.

![1686103124966](https://github.com/raphaelvallat/yasa/assets/134837471/e4597b5b-564a-4117-b9f6-afddb0e31ee0)

![1686103437495](https://github.com/raphaelvallat/yasa/assets/134837471/f16eaade-5e64-4acd-9562-07f8a95f27e0)

The above results are from the same person. We could see that  the number of so-spindle couplings I obtained on ‘F4-M1’ channel was much larger than the number of spindles. 

Where did it go wrong?

Thanks!

**Reply by raphaelvallat** [reply]
Could you re-send the second screenshot with the column names? I don't understand why you said that "was much larger than the number of spindles". Thanks

**Reply by qaqis95** [reply]
Thanks @raphaelvallat !
![1686185973372](https://github.com/raphaelvallat/yasa/assets/134837471/6487aa07-8a7f-4b66-a16a-0e9552917cb1)

Based on my understanding,  the value of '620' obtained from 'yasa.sw_detect(coupling=True)' represents the number of so-spindle couplings detected by the F4-M1 channel.

However, the number of spindles I obtained on the same channel was 96. 

It is strange that the number of couplings is even higher than the number of spindles. I believe the coupling detection should be based on spindles. I have noticed this phenomenon in many participants, I'm unsure if there's something wrong with my understanding.

**Reply by raphaelvallat** [reply]
Where do you see the 620 on the above screenshot? 

> represents the number of so-spindle couplings detected by the F4-M1 channel.

No, when you run `yasa.sw_detect(coupling=True)`, the coupling values are not based on the spindles detection. Instead, for each slow-waves, the coupling with the filtered sigma band is calculated.

**Reply by qaqis95** [reply]
Thank you so much for your thorough explanation! @raphaelvallat 
I think I have understood the 'yasa.sw_detect(coupling=True)' .
  1. The value 620 represents the number of rows in the screenshot's bottom-left corner. I think this value should correspond to the count of sigma-so couplings. Is that correct?
![1686185973372](https://github.com/raphaelvallat/yasa/assets/134837471/4ea599e5-c291-453a-bd31-c0a1bd22b451)

  2. Regarding the so-spindle couplings, what method would you recommend? I also noticed that 'find_cooccurring_spindles' does not seem to provide information on the phase and strength of the so-spindle coupling.
  
  3. Currently, I have completed data analysis for 20 participants by Yasa. The sleep stages were manually classified. One of the participants had 126 minutes of N2 sleep and 30 minutes of N3 sleep, but very few or even no spindles were detected. What do you think could be the possible reasons for this? 
![1686232674975](https://github.com/raphaelvallat/yasa/assets/134837471/061c05d0-fe31-4932-881b-48487dc1979e)

Thanks!

**Reply by raphaelvallat** [reply]
Hi,

1. No, that value represents the total number of detected slow-waves. Spindles are not detected here.
2. I do not recommend using the find_cooccurring_spindles to calculate coupling. Please see this notebook for more details: https://github.com/raphaelvallat/yasa/blob/master/notebooks/12_SO-sigma_coupling.ipynb
3. Have you check manually that you see spindles in the signal? Could the amplitude of the signal be wrong, or noisy? You could try running the spindles detection with only 2 of the three thresholds to see if you get more detected spindles.

**Reply by qaqis95** [reply]
Thank you so much for your reply @raphaelvallat . I think this will be my final question. 

The value of 620 was obtained through the 'sw = yasa.sw_detect(raw_eeg, sf, ch_names=[], hypno=hypno, include=(2, 3), coupling=True, coupling_params=dict(freq_sp=(12, 16), time=2, p=None))'. 

Do you think this value (620) represents the total number of slow waves or the total number of slow waves coupled with sigma? I'm not quite sure about this, as I think the total number of slow waves may be obtained through 'sw = yasa.sw_detect(raw_eeg, amp_ptp=(50, 400))'.
![1686406905651](https://github.com/raphaelvallat/yasa/assets/134837471/40c1eb5a-6c31-4a40-95db-f4d8f65cd15a)

The number of So obtained from these two functions differs significantly, so I still believe that 620 represents the number of SO-sigma coupling. 

I apologize for any foolish questions I may have asked.

**Reply by raphaelvallat** [reply]
620 represents the total number of slow-waves. It is expected that the number will change if you manually change the detection thresholds by using: `amp_ptp=(50, 400))`. You can get the total number of slow-waves with some significant coupling with the sigma band by doing `(sw_results["ndPAC"] > 0).sum()`. However this will only work with the default `coupling_params=dict(p=0.05)` and not `p=None`.

Hope this helps,
Thanks
